irodsadmin/management/commands/irodsorphans.py

```diff
@@ -24,184 +25,175 @@
+            if (
+                self._is_zone(collection)
+                or self._is_assay_or_study(collection)
+                or self._is_project(collection)
             ):
                 if collection.path not in expected:
-                    orphans.append(collection.path)
-    return orphans
-
-
-def get_output(orphans, irods_backend, irods):
-    lines = []
-    for orphan in orphans:
-        stats = irods_backend.get_object_stats(irods, orphan)
-        m = re.search(r'/projects/([^/]{2})/(\1[^/]+)', orphan)
+                    self._write_orphan(collection.path, irods)
+
+        for assay in assays:
+            if not assay.get_plugin():
+                continue
+            with self.irods_backend.get_session() as irods:
+                for collection in self.irods_backend.get_child_colls(
+                    irods, self.irods_backend.get_path(assay)
+                ):
+                    if collection.path not in expected:
+                        self._write_orphan(collection.path, irods)
+
+    def _write_orphan(self, path, irods):
+        stats = self.irods_backend.get_object_stats(irods, path)
+        m = re.search(r'/projects/([^/]{2})/(\1[^/]+)', path)
         if m:
             uuid = m.group(2)
             try:
                 project = Project.objects.get(sodar_uuid=uuid)
                 title = project.full_title
             except Project.DoesNotExist:
-                title = '<DELETED>'
+                title = DELETED
         else:
-            uuid = '<ERROR>'
-            title = '<ERROR>'
-        lines.append(
+            uuid = ERROR
+            title = ERROR
+        sys.stdout.write(
             ';'.join(
                 [
                     uuid,
                     title,
-                    orphan,
+                    path,
                     str(stats['file_count']),
                     filesizeformat(stats['total_size']).replace(u'\xa0', ' '),
                 ]
             )
+            + '\n'
         )
-    return lines
-
-
-class Command(BaseCommand):
-    """Command to find orphans in iRODS collections."""
-
-    help = 'Find orphans in iRODS project collections.'
 
     def handle(self, *args, **options):
-        irods_backend = get_backend_api('omics_irods')
         studies = list(Study.objects.all())
         assays = list(Assay.objects.all())
         expected = (
-            *get_assay_collections(assays, irods_backend),
-            *get_study_collections(studies, irods_backend),
-            *get_zone_collections(irods_backend),
-            *get_project_collections(irods_backend),
-            *get_assay_subcollections(studies, irods_backend),
+            *self._get_assay_collections(assays),
+            *self._get_study_collections(studies),
+            *self._get_zone_collections(),
+            *self._get_project_collections(),
+            *self._get_assay_subcollections(studies),
```

Comment:
While you were on vacation, I added issue #1741 with the suggestion it would be done together with this task.

This will require some reactoring:

- Iterate through projects (sorted by `full_title`)
- For each project, get expected collections
- For maximum clarity, for each project sort expected collections within that project before adding to the list

This should ensure the results are displayed in an order sorted by project. Tests should be updated to ensure this of course.

Reply:
Hmm, I don't seem to get the point. I pushed through the implementation where we sort projects by full title, but I'm not sure if that's exactly what you asked for.

In the current implementation, the orphan registration ordering does not depend on the `expected` list, but on the `collections`, which we get with `irods.collections.get('/{}/projects'.format(irods.zone) )`. If we want to change the order of the output, then we must filter this list of `collections`, which is not a trivial task since there are deleted projects (without full_title) or assays that are linked to projects through studies and investigations.

Could you clarify what exactly is needed, because I'm afraid I didn't get the point of the task...

Reply:
Yes, you are correct that the current implementation does not iterate by project but by the iRODS collection and I would like to change that. Right now, you can have the same project referred to in the output twice in different locations, due to first checking for orphaned landing zones, then for orphaned assay collections, etc.

A possible simple fix for this would be the following:

- After querying for all project collections with `get_colls_recursively()` in `_get_orphans()`, sort them by full path
    * This will already sort them by project, although not necessarily in the correct order by full project title
- How to make this even nicer?
    * Query for all projects of type PROJECT, store in dict as `{sodar_uuid: full_title}`
    * Compare iRODS paths to project UUID
    * Sort collections by project full_title based on the dict
    * Then continue onwards with evaluating collections

At least the first step is definitely needed, but the second one would be super nice to have, as it eliminates the manual work of sorting the output in e.g. Excel.